### ISSUE TITLE
Hsp class is not typesafe

### ISSUE THREAD

**Issue by sbliven** [biojava/biojava#354]
The Hsp class and classes that use it (e.g. BlastXMLParserTest) give lots of 'Raw type' warnings. In this case these are not benign warnings, but flag a real possibility for some type errors. The problem is that we're trying to instantiate a sequence from a string without knowing what type of sequence that is. For instance, if you built an Hsp<ProteinSequence, AminoAcidCompound> with a string like "ACTG" then it would actually create a DNASequence, potentially causing errors.

I'm not sure what the best pattern to avoid this is. Maybe the constructor to Hsp needs to take a pair of SequenceReader objects, which can be supplied by the calling code (e.g. BlastHspBuilder).

Original author: @paolopavan

**Comment by paolopavan** [comment]
Hi Spencer, 
I guess you refer to the getSequence() method. The point is that this information comes from a text report so we have to infer somehow if we are in context of DNA or proteins. This can happen at HSP class level as in this implementation or in a different one, but will still be an inference. To make it more accurate in this specific case I could exploit some information in the blast report or use the setQueryReferences()/setDatabaseReferences() mechanism but it is not guaranteed that there is such information in the generic search report or in the second case, the user had the possibility to use that API.
Do I have got the sese?

**Comment by sbliven** [comment]
The issue is not that we have to guess what type of Sequence we have; that is unavoidable. The problem is that *the type guessed by getSequence() is not neccessarily the type of the Hsp returned*. Ideally the logic would go something like "if( T extends DNASequence && string looks like DNA) {new DNASequence(string)} else if( T extends RNASequence ...." However, due to type erasure there is no way to get any information about T at runtime.

All the work-arounds I can think of would require to changing the API.

**Comment by sbliven** [comment]
I took a stab at making our sequence model type safe and found myself spiraling into insanity. There are over 100 raw type or type safety warnings in core only. Fixing all of these is daunting.

**Comment by paolopavan** [comment]
Hi Spencer, I hope to have managed the thing, I made some modifications to allow better type checking without changing the API: the sequence type now can be declared at parsing time specifying the class of the sequence using in the HSP builder. Also the mechanism of generics extended to the HSP guarantees **at API side** that the SequencePair obtained from getAlignment() method will be of the same type of the HSP itself, some tests cases are also provided to check this. 

Anyway, as I told above, choosing the right sequence type it is always a matter of guess and even with the use of generic it is not guaranteed that the guessed type is of the correct type because we do not start with a Sequence Object that can be used to instantiate the resulting ones. In the specific case of BlastXMLParser I have exploit a report feature that should be able to guarantee the correctness of the choice, but with other parsers this should not be the case. 

Beside this, may I ask what is the design problem that brought to choose this use of generics in the Sequence framework? It may well be the reason of so many type safe warnings in core. Dealing with generics has this known side effect and in our data structure we have two! We should reflect upon this, because this API is really hard to use.
Also I have a doubt but I may miss something: in the Sequence<S,C> where S extends Sequence why C extends Compound? It should be more useful if it would extend CompoundSet instead?

**Comment by sbliven** [comment]
@paolopavan I think you're on the right track, although I would use a `Class<S> sequenceClass` rather than storing an instance (similar to your 3f88b19, but with the `<S>` to ensure type safety). I made my own SearchIO_branch to illustrate.

However, this has a problem. Although it is possible to create Hsp instance with custom sequence types, there is no way to actually use such a thing, since it would throw an error (my branch) or generate logging messages and return null (your branch) for sequences that aren't DNA, RNA, or Protein (or dangerously, things with a String constructor). So it's not really as *generic* as the type declaration indicates (no pun intended).

Additionally, these changes move the problem upstream. BlastXMLParser is totally not type-safe, and so it will have to do some reflection or something to figure out the correct type of Hsp to create (as well as Hits, etc). I think we're on the right track, but it's a tangle right now.

>Beside this, may I ask what is the design problem that brought to choose this use of generics in the Sequence framework?

Regarding the choice to use generics, this was a major motivation for writing BioJava 3. Sure, all the warnings that come from generics can be a pain, but they generally indicate that you weren't thinking carefully enough about the types of your objects.

In this case, if you want to write a method to convert arbitrary Strings into Sequences, then of course that requires guessing the type of the sequence. You should get warnings there, and you can safely suppress them because the whole point of the method is to determine the type. But once you have a Sequence S, it is very helpful to be guaranteed that all operations will return that S. 

> Also I have a doubt but I may miss something: in the Sequence where S extends Sequence why C extends Compound? It should be more useful if it would extend CompoundSet instead?

A sequence has to know what compounds it is composed of, but it doesn't care about other possible compounds. For instance, a sequence is DNA whether it comes from a DNACompoundSet or a AmbiguityDNACompoundSet.

**Comment by sbliven** [comment]
So, I spent some time looking at this, but haven't quite nailed it. I started by adding generic types for Hit, Result, BlastHit, BlastResult, etc. This got me nowhere, since it doesn't solve the fundamental problem of when types are known. I think the proper approach is to start with BlastXMLParser and decide at which point the type of the result gets determined.

For instance, createObjects can look at the strings and determine the desired type before the result gets created. Then, the results have a known, well-defined type, and the list of results has type `List<Result<?,?>>`. I think this approach would work, but I haven't gotten it compiling yet.

A few questions:
* Can the results from createObjects be from different types of sequences?
* Is there downstream code which needs to work differently for different Sequence subclasses? If so then it might be a problem, since they would just have a `Result<? extends Sequence, ? extends Compound>`.

**Comment by paolopavan** [comment]
Hi Spencer,
I have some ideas on this topic but I haven't still got time to work on it. 
But I can move forward in one week or so. 
I'll post the results as well I will answer accurately to your questions. 

To anticipate something, at designing time I didn't see the need to extend sequence generics to hit and results since the only class that deal with Sequence is Hsp, which is aware of the aligned portions of the query and hit sequences. This should be enough to answer "No" to both your questions. 
Bye bye!

**Comment by josemduarte** [comment]
Any news on this? It would be good to have this for the 4.2 release that's coming soon.

**Comment by paolopavan** [comment]
Hi Jose and Spencer,
Yes, there there are news about this. There fact is that generics in Java are handled by erasure mechanism, so they are used at compile time to ensure that the code is congruent and then they are removed in the resulting bytecode. 

So, there is no possibility to know at runtime the data type of S extends Sequence which would be useful to implement a switch or a casting (see [Hsp.java#142/144/146] (https://github.com/paolopavan/biojava/blob/SearchIO_branch/biojava-core/src/main/java/org/biojava/nbio/core/search/io/Hsp.java#L142)).

Also since the signature is `<Sequence,Compound>`, even using `Class<S> sequenceClass` in [Hsp.java#274](https://github.com/paolopavan/biojava/blob/SearchIO_branch/biojava-core/src/main/java/org/biojava/nbio/core/search/io/Hsp.java#L274) the resulting runtime type is `Class<Sequence>` and cannot be explicitly converted in any of `Class<DNASequence>` or `Class<ProteinSequence>`.

Anyway I spent some time searching for a workaround. I have tried to create two types of Hsp (BlastNucleotideHsp e BlastProteinHsp) which override the property `Class<DNASequence> sequenceClass` and `Class<ProteinSequence> sequenceClass` , see [BlastXMLParser.java#261/269](https://github.com/paolopavan/biojava/blob/SearchIO_branch/biojava-core/src/main/java/org/biojava/nbio/core/search/io/blast/BlastXMLParser.java#L261).
Also extending typing to BlastXMLParser is useless because somewhere it should be set `Class<S> sequenceClass = DNASequence.class / ProteinSequence.class` which is illegal.

So said, the safety of the code may be implemented by means of business logic. There is already a code snip that uses a feature in the report that guarantee the result. It may be enforced also the API at setQueryReferences() method that could set the expected Hsp type.

*To summarize, I'm not sure this could be considered a bug since we must accept the possibility that information in the report may not be congruent with those expected. In my opinion the code can be cleaned to a couple of commit ago because generics cannot guarantee that the sequence read in the report is of the expected type. If it should still be enforced, in my opinion is sure that it will be done by some business logic check*
What do you think?

**Comment by sbliven** [comment]
I've been working on this again, and am hoping to include it in 5.0. See sbliven/biojava-sbliven/tree/fix354.

I've also asked a question on SO about how to best deal with the uncertainty of the Sequence type: http://stackoverflow.com/questions/40598972/java-two-level-parameterized-type-inference

**Comment by paolopavan** [comment]
Thank you Spencer for pushing this up, I am following with attention your
post on SO and I’m available if some more work is needed to complete this.
I may be wrong, but I feel that the problem is that we are doing with
generics a job that would have been done more efficiently in some way by
inheritance. If we guess for a protein we instantiate a ProteinSequence or
a DNA otherwise. I also have the feeling that any solution that could come
out from this would be a dirty workaround and not a neat and clean
solution. I hope to be wrong.

Spencer, by the way, please remember that the current biojava master does
not include the last improvement that I made to the code (bugfix and tests
included). They are available in my branch (see for example the links I
have mentioned in my posts). As written, a good checkpoint is a couple of
commits ago (for example df6cf5811d6c6fd2866aca8410c9915f85439f8a), those
last are there to show that even passing the Sequence class does not solve
the problem.

2016-11-15 11:46 GMT+01:00 Spencer Bliven <notifications@github.com>:

**Comment by sbliven** [comment]
We now have three competing pull requests for this issue, none of which are particularly satisfactory.

1. #607 Parameterizes everything. This is unsatisfactory because it can't be used from SPI, it's hard to use without already knowing the type of the sequences at compile time, and generics are a pain in the butt.
2. #608 Drop the whole package. But the package is providing real value (converting text to biojava objects) and is being actively used.
3. #618 Partial parameterization. This feels inconsistent.

At this point maybe we should just decide to reject all three PRs and just live with the type warnings.

**Comment by paolopavan** [comment]
Thank you Spencer for reviewing. Indeed you are right. Anyway for completeness it should be said that the complexities arise somewhere else upstream and this issue can also be found in other different parts of biojava.

Anyway, I will keep working on it as soon as I will have the possibility.

24 mar 2017, 11:26 +0100, Spencer Bliven <notifications@github.com>, ha scritto:

**Comment by lafita** [comment]
If we reject the three PRs and leave the type warnings, can we add some logger warnings to notify the user of the possible errors? 
That way, if the users are unaware of them, they can review the code and make sure everything is correct or ignore them if they know what they are doing.
Also, if an unexpected error occurs, the warnings might be a good hint to where the bug is.

**Comment by lafita** [comment]
What I meant is simply to add a log warning message every time there is a risk to receive these type of runtime exceptions (because of the code is unsafe). Even though they cannot be avoided, the user will get a clue to what is happening:

<pre>
WARN   Unsafe operation: risk to get a runtime exception if <a reason here>. See <documentation reference>
ERROR Exception...
</pre>

**Comment by paolopavan** [comment]
I might prefer to use the exception handling mechanism in this case instead. But anyway I will consider it.

29 mar 2017, 13:55 +0200, Aleix Lafita <notifications@github.com>, ha scritto:

**Comment by sbliven** [comment]
I agree with @paolopavan that runtime exceptions are the best way to handle this.

**Comment by sbliven** [comment]
I'm going to close this PR as well as the other attempts at making the search.io package type-safe.

**Comment by paolopavan** [comment]
Ok Spencer. Anyway I will still work on it as soon as my company will allow me a time slot. There is interest, but I have to cope with other stuffs also. Will I open a different issue or continue on this one?

**Comment by sbliven** [comment]
@paolopavan You can reopen this issue if you make some progress. I think the package should stay as it is (runtime errors) for the 5.0.0 release. Perhaps we should mark the API as experimental/likely to change somehow.
